tsl/src/compression/compression.c

```diff
@@ -2340,42 +2484,52 @@ decompress_batches(Chunk *ch, ScanKeyData *scankeys, int num_scankeys, Bitmapset
  *  4. Update catalog table to change status of moved chunk.
  */
 void
-decompress_batches_for_update_delete(List *chunks, List *predicates)
+decompress_batches_for_update_delete(List *chunks, List *predicates, EState *estate)
 {
-	List *segmentby_columns = NIL;
-	List *segmentby_columns_value = NIL;
-	List *is_null_check = NIL;
-	List *is_null = NIL;
+	List *segmentby_predicates = NIL;
 	ListCell *ch = NULL;
 	if (predicates)
-		fill_predicate_context(linitial(chunks),
-							   predicates,
-							   &segmentby_columns,
-							   &segmentby_columns_value,
-							   &is_null_check,
-							   &is_null);
+		segmentby_predicates = get_segmentby_predicates(linitial(chunks), predicates);
```

Comment:
This function get_segmentby_predicates() will parse the WHERE clause and return a list containing all SEGMENT BY columns specified in WHERE.
Calling same function for every chunk is a repetitive job, thus want to optimize it.

Reply:
I think the case when multiple chunks are in action is most likely uncovered - and I would be more worried about that there are mismatches between the chunks because of column drops in-between chunk creations - than to try to optimize that case.

In fact in most cases 1 chunk will be affected.
As about optimizing this: the cost of opening a table scan I think is magnitudes higher than running this predicates processing function.
In case for some reason this method does lit up during performace profiling - we can optimize it that time...

I think an assert would be better; or place it inside the below loop ; so that it doesn't run into those issues...

Reply:
This is fixed as part of a different PR.